craft_parts/executor/executor.py

```diff
@@ -0,0 +1,197 @@
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Definitions and helpers for the action executor."""
+
+import logging
+from typing import Dict, List, Union
+
+from craft_parts import callbacks, packages, parts
+from craft_parts.actions import Action, ActionType
+from craft_parts.infos import PartInfo, ProjectInfo
+from craft_parts.parts import Part
+from craft_parts.steps import Step
+from craft_parts.utils import os_utils
+
+from .collisions import check_for_stage_collisions
+from .part_handler import PartHandler
+
+logger = logging.getLogger(__name__)
+
+
+class Executor:
+    """Execute lifecycle actions."""
+
+    def __init__(
+        self,
+        *,
+        part_list: List[Part],
+        project_info: ProjectInfo,
+        extra_build_packages: List[str] = None,
+        extra_build_snaps: List[str] = None,
+    ):
+        self._part_list = part_list
+        self._project_info = project_info
+        self._extra_build_packages = extra_build_packages
+        self._extra_build_snaps = extra_build_snaps
+        self._handler: Dict[str, PartHandler] = {}
+
+    def prologue(self) -> None:
+        """Prepare the execution environment.
+
+        This method is called before executing lifecycle actions.
+        """
+        self._install_build_packages()
+        self._install_build_snaps()
+
+        callbacks.run_prologue(self._project_info, part_list=self._part_list)
+
+    def epilogue(self) -> None:
+        """Finish and clean the execution environment.
+
+        This method is called after executing lifecycle actions.
+        """
+        callbacks.run_epilogue(self._project_info, part_list=self._part_list)
+
+    def execute(self, actions: Union[Action, List[Action]]) -> None:
+        """Execute the specified action or list of actions.
+
+        :param actions: An :class:`Action` object or list of :class:`Action`
+           objects specifying steps to execute.
+
+        :raises InvalidActionException: If the action parameters are invalid.
+        """
+        if isinstance(actions, Action):
+            actions = [actions]
+
+        for act in actions:
+            self._run_action(act)
+
+    def clean(self, initial_step: Step, *, part_names: List[str] = None) -> None:
+        """Clean the given parts, or all parts if none is specified.
+
+        :param initial_step: The step to clean. More steps may be cleaned
+            as a consequence of cleaning the initial step.
+        :param part_names: A list with names of the parts to clean. If not
+            specified, all parts will be cleaned.
+        """
+        selected_parts = parts.part_list_by_name(part_names, self._part_list)
+
+        selected_steps = [initial_step] + initial_step.next_steps()
+        selected_steps.reverse()
+
+        for part in selected_parts:
+            self._create_part_handler(part)
+            handler = self._handler[part.name]
```

Comment:
it would be simpler if `_create_part_handler` would return the handler, it makes more sense to write here:

```
handler = self._create_part_handler(part)
```

Reply:
Much better, thanks.